plugins/services/src/js/reducers/form/EnvironmentVariables.js

```diff
@@ -0,0 +1,106 @@
+import {
+  ADD_ITEM,
+  REMOVE_ITEM,
+  SET
+} from '../../../../../../src/js/constants/TransactionTypes';
+import Transaction from '../../../../../../src/js/structs/Transaction';
+
+module.exports = {
+  JSONReducer(state, {type, path, value}) {
+    if (this.env == null) {
+      this.env = [];
+    }
+
+    if (path != null && path.join('.').search('env') !== -1) {
+      if (path.join('.') === 'env') {
+        switch (type) {
+          case ADD_ITEM:
+            this.env.push({key: null, value: null});
+            break;
+          case REMOVE_ITEM:
+            this.env = this.env.filter((item, index) => {
+              return index !== value;
+            });
+            if (this.env.length === 0) {
+              this.env.push({key: null, value: null});
+            }
+            break;
+        }
+
+        return this.env.reduce((memo, item) => {
+          if (item.key != null || item.value != null) {
+            memo[item.key] = item.value;
+          }
+          return memo;
+        }, {});
+      }
+
+      const joinedPath = path.join('.');
+      const index = joinedPath.match(/\d+/)[0];
+      switch (true) {
+        case (type === SET && `env.${index}.key` === joinedPath):
+          this.env[index].key = value;
+          break;
+        case (type === SET && `env.${index}.value` === joinedPath):
+          this.env[index].value = value;
+          break;
+      }
+    }
+    return this.env.reduce((memo, item) => {
+      memo[item.key] = item.value;
+      return memo;
+    }, {});
+  },
+  JSONParser(state) {
+    if (state.env == null) {
+      return [];
+    }
+
+    return Object.keys(state.env).reduce(function (memo, key, index) {
+      memo.push(new Transaction(['env'], index, ADD_ITEM));
```

Comment:
Also we need this to add a item / row in the form. Otherwise we would have logic in the form.  which should be in the Batch. But I am open to another solution here.